### DISCUSSION TITLE
Can mockall be used to mock a fn of an external crate

### DISCUSSION THREAD

**Discussion by zephod77** [asomers/mockall#445] [Questions] [answered]
Say I have this code:
```
use zmq::*; 

pub struct S {
   zmq_ctx: zmq::Context,
}
impl S {
    pub fn init(&mut self) -> bool {
        let pub_skt = match self.zmq_ctx.socket(zmq::PUB) {
            Ok(skt) => skt, 
            Err(e) => {println!("Error creating socket: {}", e); return false;},
        };
        true
    }
}
```
and I want to test the error case where socket creation fails.
```
#[cfg(test)]
mod the_test {
    use super::*;
    use mockall::*;
    use mockall::predicate::*;

    #[automock]
    pub trait MockSocket {
    }
    pub trait MockZmqContext {
        fn socket (&self, t: i32) -> Result<MockSocket>;
    }

    #[test]
    fn init_test() {
        let mut ctx_mock = MockZmqContext::new();
        let mut pub_skt_mock = MockSocket::new();

        ctx_mock.expect_socket()
            .with(predicate::eq(zmq::PUB))
            .returning(Err("Error"));

        let mut test_obj = S { zmq_ctx: &mut ctx_mock };
        assert!(!test_obj.init());
    }
}
```
The problem, or at least one problem, is with the definition of the traits and the mock objects. Another is with my understanding of how mockall works. Does 
```
    #[automock]
    pub trait MockSocket {
    }
```
automatically create a mock of the zmq::Socket stuct? The error I am getting is with the declaration of the socket function in MockZmqContext. The compiler is telling me that I have to put 'dyn' in front of MockSocket because MockSocket is a trait object but that leads to other errors and I think the problem is more fundamental than that.  
Can anyone point me to the correct path?

**Comment by asomers** [comment, marked as answer]
Is `zmq::Socket` a struct or a trait?  Because what you wrote assumes that it's a trait.  And if you fix that , then what it will create is a struct with the same API as the real struct, but it can't be passed to functions expecting the real struct.  So you need to make the substitution at compile time.  First fix your trait/struct confusion and then the next step will be more obvious.

**Reply by zephod77** [reply]
Well I do know that zmq::Socket is a struct but I'm finding the documentation confusing. 

In the User Guide section on External traits, it says "Mockall can mock traits and **_structs_** defined in external crates that are beyond your control, but you must use mock! instead of #[automock]" and it goes on to show how to mock a Clone trait. I want to mock the zmq::Context struct so I can call the socket() function on the mocked context. 

I tried using mock! but I did not understand how to use it correctly.
```

**Reply by zephod77** [reply]
I think I'm making some progress. I tried
`mock! { zmq::Context }`
but that gives 
> error: expected curly braces

so I removed the zmq:
```
mock! {
    Socket {}
}
mock! {
    Context {
        pub fn socket(&self, t: SocketType) -> Result<MockSocket>;
    }
}
...
let mut ctx_mock = MockContext::new();
ctx_mock.expect_socket().with(predicate::eq(zmq::PUB)).returning(Err("Error"));
```
but this gives an error:

> error[E0277]: expected a `FnMut<(zmq::SocketType,)>` closure, found `std::result::Result<_, &str>`
>   --> src/main.rs:44:24
>    |
> 44 |             .returning(Err("Error"));
>    |              --------- ^^^^^^^^^^^^ expected an `FnMut<(zmq::SocketType,)>` closure, found `std::result::Result<_, &str>`
>    |              |
>    |              required by a bound introduced by this call
>    |
>    = help: the trait `FnMut<(zmq::SocketType,)>` is not implemented for `std::result::Result<_, &str>`

because there is a type definition for zmq::Result
`pub type Result<T> = Result<T, Error>`
so somehow this I have to tell mock about this type definition in the zmq crate.

**Reply by asomers** [reply]
That error is happening because you're using `returning`, which expects a closure argument.  So either give it a closure argument (`.returning(|| Err("Error"))`) or use `return_const` instead: `.return_const(Err("Error"))`.

**Reply by zephod77** [reply]
Almost there.
If I use (` .returning(|e| Err(zmq::Error::EACCES))` ), it compiles 
but if I use ( `.return_const(Err::Error::EACCES))` ), the compiler complains that 

> the trait `Clone` is not implemented for `std::result::Result<_, zmq::Error>`

Now the only remaining error is: given: 
```
pub struct S { zmq_ctx: zmq::Context };
...
let mut ctx_mock = MockContext::new();
let mut test_obj = S { zmq_ctx: &mut ctx_mock };
assert!(!test_obj.init());

```

> error[E0308]: mismatched types
>   --> src/main.rs:45:41
>    |
> 45 |         let mut test_obj = S { zmq_ctx: ctx_mock };
>    |                                         ^^^^^^^^ expected struct `zmq::Context`, found struct `MockContext`
> 
> For more information about this error, try `rustc --explain E0308`.

How do I get the mocked context into my struct?

**Reply by asomers** [reply]
You have to make the substitution at compile time.  If the mock struct were in the same module as the real one, then you could use `mockall_double`.  But it isn't, so you can't.  Instead you'll have to do it like this:
```rust
#[cfg(not(test))]
use zmq::Context;
#[cfg(test)] 
use MockContext as Context;
```

**Reply by zephod77** [reply]
Whew! Made it work. I'm putting the full example here for reference.
@asomers, Thanks for the help and please let me know if there are improvements that can be made.
```
#[cfg(not(test))]
use zmq::Context as MyZmqContext;
#[cfg(test)]
use the_test::MockContext as MyZmqContext;

pub struct S {
   zmq_ctx: MyZmqContext,
}
impl S {
    pub fn init(&mut self) -> bool {
        let _pub_skt = match self.zmq_ctx.socket(zmq::PUB) {
            Ok(skt) => {println!("Successfully created socket"); skt},
            Err(e) => {println!("Error creating socket: {}", e); return false;},
        };
        true
    }
}

fn main() {
    let mut s = S { zmq_ctx: MyZmqContext::new() };
    assert!(s.init());
}

#[cfg(test)]
mod the_test {
    use super::*;
    use mockall::{*, predicate::*};

    mock! {
        pub Socket {}
    }
    mock! {
        pub Context {
            pub fn socket(&self, t: zmq::SocketType) -> Result<MockSocket, zmq::Error>;
        }
    }

    #[test]
    fn init_test() {
        let mut ctx_mock = MockContext::new();

        ctx_mock.expect_socket()
            .with(predicate::eq(zmq::PUB))
            .returning(|_e| Err(zmq::Error::EACCES));

        let mut test_obj = S { zmq_ctx: ctx_mock };
        assert!(!test_obj.init());
    }
}
```
$ cargo run
   Compiling junk v0.1.0 (/home/developer/eclipse-workspace/rust/junk)
    Finished dev [unoptimized + debuginfo] target(s) in 0.31s
     Running `target/debug/junk`
Successfully created socket

$ cargo test -- --nocapture
   Compiling junk v0.1.0 (/home/developer/eclipse-workspace/rust/junk)
    Finished test [unoptimized + debuginfo] target(s) in 0.50s
     Running unittests src/main.rs (target/debug/deps/junk-c39160b0e41ff66e)

running 1 test
Error creating socket: Permission denied
test the_test::init_test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

**Reply by asomers** [reply]
Glad to help.  Could you please select "Mark as Answer" for this thread?

**Reply by zephod77** [reply]
I have extended my example and I have more questions if you'll indulge me.
I added:
`if !pub_skt.bind(PUB_ADDR).is_ok() {...}`
to my init() function and then I added 
`pub fn bind(&self, addr: &str) -> Result<(), zmq::Error>;`
to the mock of Socket and changed the test like this:
```
let mut pub_skt_mock = MockSocket::new();
ctx_mock.expect_socket()
.with(predicate::eq(zmq::PUB))
.returning(|_| Ok(pub_skt_mock));
pub_skt_mock.expect_bind()
.with(predicate::eq(PUB_ADDR)
.returning(|_| Ok(()));
```
but this results in this error:

> error[E0507]: cannot move out of `pub_skt_mock`, a captured variable in an `FnMut` closure
>   --> src/main.rs:74:31
>    |
> 67 |         let mut pub_skt_mock = MockSocket::new();
>    |             ---------------- captured outer variable
> ...
> 74 |             .returning(|_| Ok(pub_skt_mock));
>    |                        ---    ^^^^^^^^^^^^ move occurs because `pub_skt_mock` has type `the_test::MockSocket`, which does not implement the `Copy` trait
>    |                        |
>    |                        captured by this `FnMut` closure
> 
> error[E0382]: borrow of moved value: `pub_skt_mock`
>   --> src/main.rs:78:9
>    |
> 67 |         let mut pub_skt_mock = MockSocket::new();
>    |             ---------------- move occurs because `pub_skt_mock` has type `the_test::MockSocket`, which does not implement the `Copy` trait
> ...
> 74 |             .returning(|_| Ok(pub_skt_mock));
>    |                        ---    ------------ variable moved due to use in closure
>    |                        |
>    |                        value moved into closure here
> ...
> 78 |         pub_skt_mock.expect_bind()
>    | 

pub_skt_mock is not Clonable or Copyable so do I need to do something with lifetimes?

**Reply by asomers** [reply]
The expectations you wrote say that `MockSocket::socket()` should return itself, similarly to `clone`.  I don't think that's what you want.  You probably want `socket()` to return a newly created mock object, right?

**Reply by zephod77** [reply]
That's what I started with but I was getting this error:

> thread 'the_test::init_test' panicked at 'MockSocket::bind(?): No matching expectation found', src/main.rs:49:5

Line 49 is the start of the mock! for Socket that just contains the bind declaration.

**Reply by asomers** [reply]
Did you set an expectation on bind?  And not just on the original MockSocket, but also on the one that you returned from `socket`?

**Reply by zephod77** [reply]
That what I was trying to do, and obviously failing,  by making the socket call return pub_skt_mock instead of MockSocket::new().

**Reply by zephod77** [reply]
So, @asomers, do you have any other suggestions for this? I can post the full code if you would like.

**Reply by asomers** [reply]
Like I said, you need to set the expectation on the MockSocket that you return from `socket`.  Do it like this:
```rust

ctx_mock.expect_socket()
.with(predicate::eq(zmq::PUB))
.returning(|_| {
    let mut pub_skt_mock = MockSocket::new();
    pub_skt_mock.expect_bind()
    .with(predicate::eq(PUB_ADDR)
    .returning(|_| Ok(()));
    Ok(pub_skt_mock));
});

```

**Reply by zephod77** [reply]
Thanks, I would not have come up with that one my own! 
The scope of pub_skt_mock is limited to the enclosure but what if my code were to have another external call that I wanted to mock that took both pub_skt and sub_skt, `extern_fn(pub_skt, sub_skt);`? How would you write the expectation for that?

**Reply by asomers** [reply]
What do you mean by "the enclosure"?  And expeting that other function is easy:
```rust
ctx.expect_extern_fn()
    .withf(|pub_skt, sub_skt| /* validate them, if you care to */ true)
    .return_const(())
```
